internal/backend/kubernetes/backend.go

```diff
@@ -116,16 +92,71 @@ func (b *Backend) retrieveByIP(ctx context.Context, ip string) (tinkv1.Hardware,
 	}
 
 	if len(hw.Items) == 0 {
-		return tinkv1.Hardware{}, fmt.Errorf("no hardware with ip '%v'", ip)
+		return tinkv1.Hardware{}, errNotFound
 	}
 
 	if len(hw.Items) > 1 {
-		return tinkv1.Hardware{}, fmt.Errorf("multiple hardware with ip '%v'", ip)
+		return tinkv1.Hardware{}, fmt.Errorf("multiple hardware found")
 	}
 
 	return hw.Items[0], nil
 }
 
-func toEC2Instance(tinkv1.Hardware) ec2.Instance {
-	return ec2.Instance{}
+// listerClient lists Kubernetes resources using a sigs.k8s.io/controller-runtime Backend.
+type listerClient interface {
+	List(ctx context.Context, list crclient.ObjectList, opts ...crclient.ListOption) error
+}
+
+func toEC2Instance(hw tinkv1.Hardware) ec2.Instance {
+	var i ec2.Instance
+
+	if hw.Spec.Metadata.Instance != nil {
+		i.Metadata.InstanceID = hw.Spec.Metadata.Instance.ID
+		i.Metadata.Hostname = hw.Spec.Metadata.Instance.Hostname
+		i.Metadata.LocalHostname = hw.Spec.Metadata.Instance.Hostname
+		i.Metadata.Tags = hw.Spec.Metadata.Instance.Tags
+
+		if hw.Spec.Metadata.Instance.OperatingSystem != nil {
+			i.Metadata.OperatingSystem.Slug = hw.Spec.Metadata.Instance.OperatingSystem.Slug
+			i.Metadata.OperatingSystem.Distro = hw.Spec.Metadata.Instance.OperatingSystem.Distro
+			i.Metadata.OperatingSystem.Version = hw.Spec.Metadata.Instance.OperatingSystem.Version
+			i.Metadata.OperatingSystem.ImageTag = hw.Spec.Metadata.Instance.OperatingSystem.ImageTag
+		}
+
+		// Iterate over all IPs and set the first one for IPv4 and IPv6 as the values in the
```

Comment:
how are these ordered? does it depend on the backend? I believe they come back from the backend as a slice? and we dont have any ordering done in kubernetes?

Reply:
Kubernetes preserves slice ordering. The contract is whatever is first is chosen.